Content length header consolidation

td-extreme opened the pull request:
This pull request ensures that the `Content-Type` and `Content-Length` headers get set anytime a `HttpResponse` has a body.
- Made is so that when setting the body of a `HttpResponse` that the `Content-Type` must be supplied. 
- The `HttpResponse` now sets the `Content-Type` header and figures out the `Content-Length` header. 

Prior to this change, each route would be responsible for making sure it set a `Content-Type` header and had to figure out the `Content-Length` itself and set that as well.

td-extreme committed:
HttpRequest now sets Content-Length header

Instead of having handlers figure out the content length it is now done
in the setBody(byte[] body) method of HttpRequest.

This was done in order to remove duplication, and ensure that the
Content-Length header always gets set when there is a body.

td-extreme committed:
setBody method now requires a content type.

This is to ensure that a Content-Type header gets set if there is a
body. Before the handler building a response would be responsible for
making a seperate method call to setHeader. This change removes that
extra responsiblity and makes setting a Content Typer mandatory if a
body is set.

mikeknep commented on src/main/java/handlers/HandlerBadRequest.java:
```diff
@@ -5,7 +5,7 @@
     HttpResponse rtnResponse = new HttpResponse();
     rtnResponse.setResponseCode(400);
-    rtnResponse.setBody("Bad Request");
+    rtnResponse.setBody("Bad Request", "test/plain");
```
spelling `text`

mikeknep commented on src/main/java/handlers/HandlerFileNotFound.java:
```diff
@@ -5,7 +5,7 @@
     HttpResponse rtnResponse = new HttpResponse();
     rtnResponse.setResponseCode(404);
-    rtnResponse.setBody("File not Found");
+    rtnResponse.setBody("File not Found", "plain/text");
```
flip

mikeknep commented on src/main/java/handlers/HandlerPostFileContents.java:
```diff
@@ -22,7 +22,7 @@ public HttpResponse generateResponse() {
     catch (IOException e) {
       e.printStackTrace();
-      rtnResponse.setBody("IOException");
+      rtnResponse.setBody("IOException", "text/plain");
```
maybe this should be defined in a constant somewhere?

edit: "this" = all these `"text/plain"` strings

mikeknep commented on src/main/java/response/ContentTypeForFileExtension.java:
```diff
@@ -1,20 +1,16 @@
-    rtnHeaders.put("Content-Length", contentLength(body));
-    return rtnHeaders;
+  public String getContentType(String fileName) {
+    return contentTypeMap.getOrDefault(getFileExtension(fileName), "application/force-download");
```
also constantize

td-extreme committed:
fixed typos

mikeknep commented on src/main/java/response/HttpResponse.java:
```diff
@@ -23,12 +23,14 @@ public int responseCode() {
+  public void setBody(byte[] body, String contentType) {
     this.body = body;
+    addHeader("Content-Length", contentLength(this.body));
+    addHeader("Content-Type", contentType);
```
👍 nice to see these here

dadarek commented on src/main/java/response/HttpResponse.java:
```diff
@@ -23,12 +23,14 @@ public int responseCode() {
   }
 
-  public void setBody(byte[] body) {
+  public void setBody(byte[] body, String contentType) {
```
I can see the `Content-Length` getting set here–nice job–_but_ I feel like the `contentType` is out of place here. Feel like it should be a separate setter altogether.

dadarek commented on src/main/java/handlers/HandlerGetDirectoryContents.java:
```diff
@@ -18,20 +18,13 @@ public HttpResponse generateResponse() {
     catch (IOException e) {
       e.printStackTrace();
-      rtnResponse.setBody("IOException");
+      rtnResponse.setBody("IOException", "text/plain");
```
Hmmm ... I sort of feel like this should be a 500 now that I think about it. _Something_ went wrong on the server, right? If we get to this point we don't _expect_ an exception.

Do we do 500's?

td-extreme replied:
Okay, we hadn't talked about 500's. I'll make a separate PR for this.

dadarek replied:
Don't worry about 500's. We'll create a story for it if there is time.

dadarek commented on src/main/java/handlers/HandlerGetFileContents.java:
```diff
@@ -4,25 +4,24 @@
 public class HandlerGetFileContents implements Ihandler {
   private IFileIO fileIO;
-  private ResponseHeadersForContent responseHeaders;
+  private ContentTypeForFileExtension contentType;
```
Here is an example of something that I feel probably doesn't have to get injected (referencing back to our conversation from Tuesday?)

This `ContentTypeForFileExtension` is a pretty simple class. Could probably just instantiate it directly.

Just my 2 cents.

mikeknep replied:
I agree, although for a different reason: to me, it feels more reasonable for these particular handlers to just know about `ContentTypeForFileExtension` than for the top-level `HandlerRouter` to have to know about such a minute detail and provide it to the handlers.

dadarek commented on src/test/java/HandlerGetFileContentsTest.java:
```diff
@@ -6,63 +6,63 @@
     assertEquals(response.headers().get("Content-Type"), "text/plain");
   }
 
   public void testContentTypeIsImageJpegForJpg() {
```
Since you have the content-type mapping already tested in `ContentTypeForFileExtensionTest`, it's _probably_ safe to rip out a couple of these tests here in `HandlerGetFileContentsTest`.

dadarek commented on src/test/java/HttpResponseWriterTest.java:
```diff
@@ -15,12 +15,12 @@ protected void setUp() {
     response.addHeader("Test/Header", "test value");
     response.addHeader("Test/Header2", "test value2");
-    String responseShouldBe = "HTTP/1.1 200 OK\r\nTest/Header: test value\r\nTest/Header2: test value2\r\n\r\nThis is the response body";
+    String responseShouldBe = "HTTP/1.1 200 OK\r\nTest/Header: test value\r\nTest/Header2: test value2\r\nContent-Length: 25\r\nContent-Type: text/plain\r\n\r\nThis is the response body";
```
Anyway to clean this one line up _juuust a little_ ?

dadarek commented:
Other than the minor nit-pics, 👍 

ship it

mikeknep commented:
Agreed with @dadarek—some small things and then ship. If you're still working on a solution for constants feel free to put that in a separate PR.

td-extreme committed:
Added Constants for Content Type

td-extreme committed:
Remove duplicate tests.

The ContentTypeForFileExtensionTest now tests that the correct content
type is selected for each type of file extenstion. So there is no need
to test every type again in HandlerGetFileContentsTest.

td-extreme committed:
Shortened generated response.

Removed the additional headers that were added, now that content type
and length are generated for this test.

td-extreme committed:
Remove injection for helper class.

ContentTypeForFileExtenstion is no a hard dependency instead of being
injected.
